rcl/src/rcl/arguments.c

```diff
@@ -0,0 +1,377 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/arguments.h"
+
+#include <string.h>
+
+#include "rcl/error_handling.h"
+#include "rcl/validate_topic_name.h"
+#include "rcutils/logging_macros.h"
+#include "rcutils/strdup.h"
+#include "./arguments_impl.h"
+#include "./remap_impl.h"
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+// instance of global arguments
+rcl_arguments_t __rcl_arguments;
+
+/// \return true if c is in [a-zA-Z0-9_]
+/// \internal
+RCL_LOCAL
+bool
+_rcl_valid_token_char(char c)
+{
+  // assumes ASCII
+  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_';
+}
+
+/// Parse an argument that may or may not be a remap rule
+/// \param[in] arg the argument to parse
+/// \param[in] allocator an allocator to use
+/// \param[in,out] output_rule input a zero intialized rule, output a fully initialized one
+/// \return RCL_RET_OK if a valid rule was parsed
+/// \return RLC_RET_ERROR if no valid rule was parsed
+/// \internal
+RCL_LOCAL
+rcl_ret_t
+_rcl_parse_remap_rule(
+  const char * arg,
+  rcl_allocator_t allocator,
+  rcl_remap_t * output_rule)
+{
+  size_t len_node_name = 0;
+  size_t len_match = 0;
+  size_t len_replacement = 0;
+
+  const char * separator = NULL;
+  const char * colon = NULL;
+  const char * match_begin = arg;
+  const char * replacement_begin = NULL;
+
+  // A valid rule has two parts separated by :=
+  separator = strstr(arg, ":=");
+  if (NULL == separator) {
+    return RCL_RET_ERROR;
```

Comment:
Improved error messages and return codes in 12e5f599606b9362dd4ce6d4ec4bdf9979f61606. I made the return code generic because this function non-public and only used in one place and I'm not sure if I'll need to change error messages when implementing the complete remapping grammar.